File: bindings/csharp/Micron.cs
// Copyright Quad4 2026
// SPDX-License-Identifier: 0BSD

using System.Runtime.InteropServices;
using System.Text.Json;

namespace Quad4.Micron;

public static class Micron
{
    private static readonly IntPtr LibHandle = LoadLibrary();

    private static string LibFileName()
    {
        if (OperatingSystem.IsWindows()) return "libmicron.dll";
        if (OperatingSystem.IsMacOS()) return "libmicron.dylib";
        return "libmicron.so";
    }

    private static IntPtr LoadLibrary()
    {
        var env = Environment.GetEnvironmentVariable("MICRON_LIB_PATH");
        if (!string.IsNullOrWhiteSpace(env) && File.Exists(env))
        {
            return NativeLibrary.Load(env);
        }

        var name = LibFileName();
        var rid = RuntimeInformation.RuntimeIdentifier;
        var candidates = new[]
        {
            Path.Combine(AppContext.BaseDirectory, "runtimes", rid, "native", name),
            Path.Combine(AppContext.BaseDirectory, "native", name),
            Path.Combine("dist", name),
            Path.GetFullPath(Path.Combine("dist", name)),
        };
        foreach (var path in candidates)
        {
            if (File.Exists(path))
            {
                return NativeLibrary.Load(path);
            }
        }
        return NativeLibrary.Load(name);
    }

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate IntPtr ConvertDelegate(IntPtr markup, int dark, int mono);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate IntPtr ParseHeaderDelegate(IntPtr markup);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate IntPtr CollectFieldsDelegate(IntPtr inputsJson);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate IntPtr BuildPayloadDelegate(IntPtr fieldsJson, IntPtr destination, IntPtr fieldsSpec);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void FreeDelegate(IntPtr ptr);

    private static readonly ConvertDelegate ConvertFn =
        Marshal.GetDelegateForFunctionPointer<ConvertDelegate>(NativeLibrary.GetExport(LibHandle, "micron_convert"));
    private static readonly ParseHeaderDelegate ParseHeaderFn =
        Marshal.GetDelegateForFunctionPointer<ParseHeaderDelegate>(NativeLibrary.GetExport(LibHandle, "micron_parse_header_tags"));
    private static readonly CollectFieldsDelegate CollectFieldsFn =
        Marshal.GetDelegateForFunctionPointer<CollectFieldsDelegate>(NativeLibrary.GetExport(LibHandle, "micron_collect_form_fields"));
    private static readonly BuildPayloadDelegate BuildPayloadFn =
        Marshal.GetDelegateForFunctionPointer<BuildPayloadDelegate>(NativeLibrary.GetExport(LibHandle, "micron_build_request_payload"));
    private static readonly FreeDelegate FreeFn =
        Marshal.GetDelegateForFunctionPointer<FreeDelegate>(NativeLibrary.GetExport(LibHandle, "micron_free"));

    private static IntPtr AllocUtf8(string s)
    {
        return Marshal.StringToCoTaskMemUTF8(s ?? "");
    }

    private static string TakeString(IntPtr ptr)
    {
        if (ptr == IntPtr.Zero) return "";
        try
        {
            return Marshal.PtrToStringUTF8(ptr) ?? "";
        }
        finally
        {
            FreeFn(ptr);
        }
    }

    public static string Convert(string markup, bool darkTheme = true, bool forceMonospace = true)
    {
        var p = AllocUtf8(markup);
        try
        {
            return TakeString(ConvertFn(p, darkTheme ? 1 : 0, forceMonospace ? 1 : 0));
        }
        finally
        {
            Marshal.FreeCoTaskMem(p);
        }
    }

    public static JsonElement ParseHeaderTags(string markup)
    {
        var p = AllocUtf8(markup);
        try
        {
            return JsonDocument.Parse(TakeString(ParseHeaderFn(p))).RootElement.Clone();
        }
        finally
        {
            Marshal.FreeCoTaskMem(p);
        }
    }

    public static JsonElement CollectFormFields(string inputsJson)
    {
        var p = AllocUtf8(inputsJson);
        try
        {
            return JsonDocument.Parse(TakeString(CollectFieldsFn(p))).RootElement.Clone();
        }
        finally
        {
            Marshal.FreeCoTaskMem(p);
        }
    }

    public static JsonElement BuildRequestPayload(string fieldsJson, string destination, string fieldsSpec)
    {
        var f = AllocUtf8(fieldsJson);
        var d = AllocUtf8(destination);
        var s = AllocUtf8(fieldsSpec);
        try
        {
            return JsonDocument.Parse(TakeString(BuildPayloadFn(f, d, s))).RootElement.Clone();
        }
        finally
        {
            Marshal.FreeCoTaskMem(f);
            Marshal.FreeCoTaskMem(d);
            Marshal.FreeCoTaskMem(s);
        }
    }
}

File: bindings/dart/lib/micron.dart
// Copyright Quad4 2026
// SPDX-License-Identifier: 0BSD

import 'dart:convert';
import 'dart:ffi';
import 'dart:io';

import 'package:ffi/ffi.dart';

typedef _ConvertNative = Pointer<Utf8> Function(Pointer<Utf8>, Int32, Int32);
typedef _ConvertDart = Pointer<Utf8> Function(Pointer<Utf8>, int, int);

typedef _ParseHeaderNative = Pointer<Utf8> Function(Pointer<Utf8>);
typedef _ParseHeaderDart = Pointer<Utf8> Function(Pointer<Utf8>);

typedef _CollectFieldsNative = Pointer<Utf8> Function(Pointer<Utf8>);
typedef _CollectFieldsDart = Pointer<Utf8> Function(Pointer<Utf8>);

typedef _BuildPayloadNative = Pointer<Utf8> Function(
  Pointer<Utf8>,
  Pointer<Utf8>,
  Pointer<Utf8>,
);
typedef _BuildPayloadDart = Pointer<Utf8> Function(
  Pointer<Utf8>,
  Pointer<Utf8>,
  Pointer<Utf8>,
);

typedef _FreeNative = Void Function(Pointer<Utf8>);
typedef _FreeDart = void Function(Pointer<Utf8>);

String _libName() {
  if (Platform.isMacOS) return 'libmicron.dylib';
  if (Platform.isWindows) return 'libmicron.dll';
  return 'libmicron.so';
}

DynamicLibrary _openLib() {
  final env = Platform.environment['MICRON_LIB_PATH'];
  if (env != null && env.isNotEmpty && File(env).existsSync()) {
    return DynamicLibrary.open(env);
  }
  final name = _libName();
  final scriptDir = File(Platform.script.toFilePath()).parent.parent.path;
  final candidates = <String>[
    '$scriptDir/native/$name',
    '$scriptDir/../../dist/$name',
    'native/$name',
    '../../dist/$name',
    'dist/$name',
  ];
  for (final path in candidates) {
    if (File(path).existsSync()) {
      return DynamicLibrary.open(File(path).absolute.path);
    }
  }
  throw StateError(
    'libmicron not found; set MICRON_LIB_PATH or place the shared library under bindings/dart/native or dist/',
  );
}

final DynamicLibrary _lib = _openLib();

final _ConvertDart _convertFn =
    _lib.lookupFunction<_ConvertNative, _ConvertDart>('micron_convert');
final _ParseHeaderDart _parseHeaderFn = _lib
    .lookupFunction<_ParseHeaderNative, _ParseHeaderDart>(
      'micron_parse_header_tags',
    );
final _CollectFieldsDart _collectFieldsFn = _lib
    .lookupFunction<_CollectFieldsNative, _CollectFieldsDart>(
      'micron_collect_form_fields',
    );
final _BuildPayloadDart _buildPayloadFn = _lib
    .lookupFunction<_BuildPayloadNative, _BuildPayloadDart>(
      'micron_build_request_payload',
    );
final _FreeDart _freeFn =
    _lib.lookupFunction<_FreeNative, _FreeDart>('micron_free');

String _takeString(Pointer<Utf8> ptr) {
  if (ptr == nullptr) return '';
  try {
    return ptr.toDartString();
  } finally {
    _freeFn(ptr);
  }
}

/// Convert Micron markup to an HTML fragment.
String convert(
  String markup, {
  bool darkTheme = true,
  bool forceMonospace = true,
}) {
  final p = markup.toNativeUtf8();
  try {
    return _takeString(
      _convertFn(p, darkTheme ? 1 : 0, forceMonospace ? 1 : 0),
    );
  } finally {
    malloc.free(p);
  }
}

/// Return page colors from leading #!fg= / #!bg= lines.
Map<String, String> parseHeaderTags(String markup) {
  final p = markup.toNativeUtf8();
  try {
    final data =
        jsonDecode(_takeString(_parseHeaderFn(p))) as Map<String, dynamic>;
    return {
      'fg': (data['fg'] ?? '') as String,
      'bg': (data['bg'] ?? '') as String,
    };
  } finally {
    malloc.free(p);
  }
}

/// Collect form field values from input snapshots.
Map<String, String> collectFormFields(List<Map<String, Object?>> inputs) {
  final p = jsonEncode(inputs).toNativeUtf8();
  try {
    final data =
        jsonDecode(_takeString(_collectFieldsFn(p))) as Map<String, dynamic>;
    return data.map((k, v) => MapEntry(k, '$v'));
  } finally {
    malloc.free(p);
  }
}

/// Build a Micron-style request payload.
Map<String, Object?> buildRequestPayload(
  Map<String, String> fields,
  String destination,
  String fieldsSpec,
) {
  final f = jsonEncode(fields).toNativeUtf8();
  final d = destination.toNativeUtf8();
  final s = fieldsSpec.toNativeUtf8();
  try {
    return jsonDecode(_takeString(_buildPayloadFn(f, d, s)))
        as Map<String, Object?>;
  } finally {
    malloc.free(f);
    malloc.free(d);
    malloc.free(s);
  }
}
